src/librustc_mir/interpret/validity.rs

```diff
@@ -95,126 +128,189 @@ fn path_format(path: &Vec<PathElem>) -> String {
-            ScalarMaybeUndef::Undef => return validation_failure!("undefined bytes", path),
-        };
+        // Go over all the primitive types
+        match ty.sty {
+            ty::Bool => {
+                try_validation!(value.to_bool(),
+                    scalar_format(value), path, "a boolean");
+            },
+            ty::Char => {
+                try_validation!(value.to_char(),
+                    scalar_format(value), path, "a valid unicode codepoint");
+            },
+            ty::Float(_) | ty::Int(_) | ty::Uint(_) if const_mode => {
+                // Integers/floats in CTFE: Must be scalar bits
+                try_validation!(value.to_bits(size),
+                    scalar_format(value), path, "initialized plain bits");
+            }
+            ty::Float(_) | ty::Int(_) | ty::Uint(_) | ty::RawPtr(_) => {
+                // Anything but undef goes
+                try_validation!(value.not_undef(),
+                    scalar_format(value), path, "a raw pointer");
+            },
+            ty::Ref(..) => {
+                // This is checked by the recursive reference handling, nothing to do here.
+                debug_assert!(ty.builtin_deref(true).is_some() && !ty.is_unsafe_ptr());
+            }
+            ty::FnPtr(_sig) => {
+                let ptr = try_validation!(value.to_ptr(),
+                    scalar_format(value), path, "a pointer");
+                let _fn = try_validation!(self.memory.get_fn(ptr),
+                    scalar_format(value), path, "a function pointer");
+                // TODO: Check if the signature matches
+            }
+            ty::FnDef(..) => {
+                // This is a zero-sized type with all relevant data sitting in the type.
+                // There is nothing to validate.
+            }
+            _ => bug!("Unexpected primitive type {}", ty)
+        }
+        Ok(())
+    }
 
+    /// Make sure that `value` matches the
+    fn validate_scalar_layout(
+        &self,
+        value: ScalarMaybeUndef,
+        size: Size,
+        path: &Vec<PathElem>,
+        layout: &layout::Scalar,
+    ) -> EvalResult<'tcx> {
+        trace!("validate scalar by layout: {:#?}, {:#?}, {:#?}", value, size, layout);
+        let (lo, hi) = layout.valid_range.clone().into_inner();
+        if lo == u128::min_value() && hi == u128::max_value() {
+            // Nothing to check
+            return Ok(());
+        }
+        // At least one value is excluded. Get the bits.
+        let value = try_validation!(value.not_undef(),
+            scalar_format(value), path, format!("something in the range {:?}", layout.valid_range));
         let bits = match value {
+            Scalar::Ptr(_) => {
+                // Comparing a ptr with a range is not meaningfully possible.
+                // In principle, *if* the pointer is inbonds, we could exclude NULL, but
+                // that does not seem worth it.
```

Comment:
I'm fairly sure I had a test for just this case (`enum Foo { E = 0 }` and then transmuting a pointer to the enum type)

Reply:
I made this more strict with the latest changes, could you have a look?

I found your test in `ub-enum.rs`. Bot it kept failing as expected... maybe because this is actually not considered a primitive type, but an enum, and hence it loads the discriminant and that always fails when it is a pointer?